tests/pmap_test.py

```diff
@@ -384,6 +384,50 @@ def sum_and_broadcast(x, axis):
     expected = sum_and_broadcast(sum_and_broadcast(x, 0), 1)
     self.assertAllClose(ans, expected, check_dtypes=False)
 
+  def testPsumReplicaGroups(self):
+    replicas = xla_bridge.device_count()
+    if replicas % 2 != 0:
+      raise SkipTest
+    axis_index_groups = onp.arange(replicas).reshape(
+      2, replicas // 2).tolist()
+    f = lambda x: x - lax.psum(x, 'i', axis_index_groups=axis_index_groups)
+    f = pmap(f, 'i')
+
+    shape = (replicas, 4)
+    x = onp.arange(prod(shape), dtype=onp.float32).reshape(shape)
+    def sum_helper(a):
+      return onp.broadcast_to(a.sum(0, keepdims=True),
+                              (replicas // 2, x.shape[1]))
+    expected_psum_1 = sum_helper(x[:replicas // 2])
+    expected_psum_2 = sum_helper(x[replicas // 2:])
+    expected_psum = onp.concatenate([expected_psum_1, expected_psum_2], 0)
+    expected = x - expected_psum
+
+    ans = f(x)
+    self.assertAllClose(ans, expected, check_dtypes=False)
+
+  def testNestedPmapReplicaGroups(self):
+    replicas = xla_bridge.device_count()
+    if replicas % 4 != 0:
+      raise SkipTest
+    axis_index_groups = onp.arange(replicas // 2).reshape(
+      2, replicas // 4).tolist()
+    f = lambda x: x - lax.psum(x, 'i', axis_index_groups=axis_index_groups)
+    f = pmap(pmap(f, 'i'), 'j')
+
+    shape = (2, replicas // 2, 4)
+    x = onp.arange(prod(shape), dtype=onp.float32).reshape(shape)
+    def sum_helper(a):
+      return onp.broadcast_to(a.sum(1, keepdims=True),
+                              (2, replicas // 4, x.shape[2]))
+    expected_psum_1 = sum_helper(x[:, :replicas // 4])
+    expected_psum_2 = sum_helper(x[:, replicas // 4:])
+    expected_psum = onp.concatenate([expected_psum_1, expected_psum_2], 1)
+    expected = x - expected_psum
+
+    ans = f(x)
+    self.assertAllClose(ans, expected, check_dtypes=False)
+
```

Comment:
Also if you wanna be super future proof... add test where the pmaps are "perfectly nested" :O (since once we don't use replication for perfectly nested pmaps we'll lose test coverage of that logic, which I just reviewed and is quite tricky!)

Or just wait until we write a fuzzer ;)

Reply:
Done (assuming you meant "imperfectly nested", since the others are perfectly nested).

Reply:
I did mean imperfectly, oops.